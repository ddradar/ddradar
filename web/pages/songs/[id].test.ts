import { mockNuxtImport } from '@nuxt/test-utils/runtime'
import Oruga from '@oruga-ui/oruga-next'
import { bulmaConfig } from '@oruga-ui/theme-bulma'
import { describe, expect, test, vi } from 'vitest'
import { ref } from 'vue'

import { testSongData } from '~~/../core/test/data'
import Page from '~~/pages/songs/[id].vue'
import { mountAsync } from '~~/test/test-utils'
import { createClientPrincipal } from '~~/test/test-utils-server'

const { useEasyAuthMock, useFetchMock, useRouteMock } = vi.hoisted(() => ({
  useEasyAuthMock: vi.fn(),
  useFetchMock: vi.fn(),
  useRouteMock: vi.fn(),
}))
mockNuxtImport('useEasyAuth', () => useEasyAuthMock)
mockNuxtImport('useFetch', () => useFetchMock)
mockNuxtImport('useRoute', () => useRouteMock)

vi.mock('~~/composables/useAuth')

describe('Page /songs/[id]', () => {
  const params = { id: testSongData.id }

  test('{ song: null } renders loading state', async () => {
    // Arrange
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useRoute).mockReturnValue({ params } as any)
    vi.mocked(useEasyAuth).mockResolvedValue({
      clientPrincipal: ref(createClientPrincipal('', '', false)),
    } as any)
    vi.mocked(useFetch).mockResolvedValue({ data: ref(null) } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */

    // Act
    const wrapper = await mountAsync(Page, {
      global: { plugins: [[Oruga, bulmaConfig]], stubs: { ChartInfo: true } },
    })

    // Assert
    expect(wrapper.element).toMatchSnapshot()
  })

  test('{ song: <Data> } renders song info', async () => {
    // Arrange
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useRoute).mockReturnValue({ params } as any)
    vi.mocked(useEasyAuth).mockResolvedValue({
      clientPrincipal: ref(createClientPrincipal('', '', false)),
    } as any)
    vi.mocked(useFetch).mockResolvedValue({ data: ref(testSongData) } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */

    // Act
    const wrapper = await mountAsync(Page, {
      global: { plugins: [[Oruga, bulmaConfig]], stubs: { ChartInfo: true } },
    })

    // Assert
    expect(wrapper.element).toMatchSnapshot()
  })

  test('{ isAdmin: true } renders edit button', async () => {
    // Arrange
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useRoute).mockReturnValue({ params } as any)
    vi.mocked(useEasyAuth).mockResolvedValue({
      clientPrincipal: ref(createClientPrincipal('', '', true)),
    } as any)
    vi.mocked(useFetch).mockResolvedValue({ data: ref(testSongData) } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */

    // Act
    const wrapper = await mountAsync(Page, {
      global: { plugins: [[Oruga, bulmaConfig]], stubs: { ChartInfo: true } },
    })

    // Assert
    expect(wrapper.element).toMatchSnapshot()
  })
})
