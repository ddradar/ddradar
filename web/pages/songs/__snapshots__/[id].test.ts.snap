// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Page /songs/[id] > { isAdmin: true } renders edit button 1`] = `
<section
  class="section"
>
  
  <h1
    class="title"
  >
    PARANOiA
  </h1>
  <h2
    class="subtitle"
  >
    180 / DDR 1st
  </h2>
  <h2
    class="subtitle"
  >
    BPM 180
  </h2>
  <div
    class="buttons"
  >
    <nuxtlink
      class="button is-info"
      to="/admin/song/06loOQ0DQb0DqbOibl6qO81qlIdoP9DI"
    >
      <span
        class="button-wrapper"
      >
        <span
          class="icon"
        >
          <i
            class="mdi mdi-pencil-box mdi-24px"
          />
        </span>
        <span
          class=""
        >
          
           編集 
          
        </span>
        <!--v-if-->
      </span>
    </nuxtlink>
  </div>
  <div
    class="content columns is-multiline"
  >
    
    <chart-info-stub
      chart="[object Object]"
      songid="06loOQ0DQb0DqbOibl6qO81qlIdoP9DI"
    />
    <chart-info-stub
      chart="[object Object]"
      songid="06loOQ0DQb0DqbOibl6qO81qlIdoP9DI"
    />
    
  </div>
  <div
    class="content columns is-multiline"
  >
    
    
  </div>
  
</section>
`;

exports[`Page /songs/[id] > { song: <Data> } renders song info 1`] = `
<section
  class="section"
>
  
  <h1
    class="title"
  >
    PARANOiA
  </h1>
  <h2
    class="subtitle"
  >
    180 / DDR 1st
  </h2>
  <h2
    class="subtitle"
  >
    BPM 180
  </h2>
  <!--v-if-->
  <div
    class="content columns is-multiline"
  >
    
    <chart-info-stub
      chart="[object Object]"
      songid="06loOQ0DQb0DqbOibl6qO81qlIdoP9DI"
    />
    <chart-info-stub
      chart="[object Object]"
      songid="06loOQ0DQb0DqbOibl6qO81qlIdoP9DI"
    />
    
  </div>
  <div
    class="content columns is-multiline"
  >
    
    
  </div>
  
</section>
`;

exports[`Page /songs/[id] > { song: null } renders loading state 1`] = `
<section
  class="section"
>
  
  <transition-stub
    appear="false"
    css="true"
    name="fade"
    persisted="false"
  >
    <div
      class="o-load loading o-load--fullpage is-full-page"
    >
      <div
        class="o-load__overlay loading-overlay"
      />
      
      <span
        class="icon is-spin is-medium o-load__icon icon"
      >
        <i
          class="mdi mdi-loading mdi-36px"
        />
      </span>
      
    </div>
  </transition-stub>
  <h1
    class="title"
  >
    <div
      class="b-skeleton is-animated"
    >
      <div
        class="b-skeleton-item is-rounded"
      />
    </div>
  </h1>
  <h2
    class="subtitle"
  >
    <div
      class="b-skeleton is-animated"
    >
      <div
        class="b-skeleton-item is-rounded"
      />
    </div>
  </h2>
  <h2
    class="subtitle"
  >
    <div
      class="b-skeleton is-animated"
    >
      <div
        class="b-skeleton-item is-rounded"
      />
    </div>
  </h2>
  <div
    class="content columns is-multiline"
  >
    <div
      class="b-skeleton is-animated"
    >
      <div
        class="b-skeleton-item is-rounded"
      />
      <div
        class="b-skeleton-item is-rounded"
      />
      <div
        class="b-skeleton-item is-rounded"
      />
    </div>
  </div>
  
</section>
`;
